Add garbage collection function for segments produced by partially failed imports

EricStarer opened the pull request:
Add garbage collection function for segments produced by partially failed imports

Signed-off-by: xingzhao <xing.zhao@zilliz.com>

https://github.com/milvus-io/milvus/issues/17814

/kind improvement

Commit:
Add garbage collection function for segments produced by partially failed imports

Signed-off-by: xingzhao <xing.zhao@zilliz.com>

soothing-rain commented on internal/datacoord/garbage_collector.go:
```diff
@@ -158,8 +163,28 @@ func (gc *garbageCollector) scan() {
 	drops := gc.meta.SelectSegments(func(segment *SegmentInfo) bool {
-		return segment.GetState() == commonpb.SegmentState_Dropped && !gc.segRefer.HasSegmentLock(segment.ID)
+		_, failedSegmentFlag := failedSegmentIDsSet[segment.ID]
+		if failedSegmentFlag {
```
This seems duplicated as the flag is checked in at line #187

EricStarer replied:
I will first set the segment state into dropped to make it easier to clear in the next time if the sevice is crashed.

soothing-rain replied:
That makes sense. Then how about just remove the flag check in the next line?

EricStarer replied:
ok

soothing-rain commented on internal/datacoord/garbage_collector.go:
```diff
@@ -158,8 +163,28 @@ func (gc *garbageCollector) scan() {
+			MsgType: commonpb.MsgType_GetImportFailedSegmentIDs,
+		},
+	}
+	ctx := context.Background()
```
Check the healthiness of rootcoord before sending RPC

EricStarer replied:
this may be checked at rootcoord server

soothing-rain replied:
Take a look at: https://github.com/milvus-io/milvus/blob/f0b036a35ae17ada3101eebc6a200b5f621b7e7e/internal/rootcoord/root_coord.go#L216

EricStarer replied:
this method is private and can only be call at rootcoord package？

soothing-rain replied:
OK I guess just checking `rcc != nil` for now

EricStarer replied:
ok

soothing-rain commented on internal/datacoord/garbage_collector.go:
```diff
@@ -54,21 +56,24 @@ type garbageCollector struct {
 	meta     *meta
 	segRefer *SegmentReferenceManager
 
+	rootCoordClient types.RootCoord
```
Go prefers shorter variable names, for example `rcc`.

EricStarer replied:
ok

soothing-rain commented on internal/datacoord/garbage_collector.go:
```diff
@@ -158,8 +163,28 @@ func (gc *garbageCollector) scan() {
+	if err != nil {
+		log.Error("getImportFailedSegmentIDs failed", zap.String("error", err.Error()))
+	}
+	failedSegmentIDs := rsp.SegmentIDs
```
Here and below, use "rsp.GetSegmentIDs()" instead.

soothing-rain commented on internal/rootcoord/root_coord.go:
```diff
@@ -3021,3 +3021,16 @@ func (c *Core) ListCredUsers(ctx context.Context, in *milvuspb.ListCredUsersRequ
 	}, nil
 }
+
+func (c *Core) GetImportFailedSegmentIDs(ctx context.Context, req *internalpb.GetImportFailedSegmentIDsRequest) (*internalpb.GetImportFailedSegmentIDsResponse, error) {
```
Comment

soothing-rain commented on internal/rootcoord/import_manager.go:
```diff
@@ -729,3 +729,18 @@ func taskExpired(ti *datapb.ImportTaskInfo) bool {
 		Params.RootCoordCfg.ImportTaskExpiration <= float64(time.Now().Unix()-ti.GetCreateTs())
 }
+
+func (m *importManager) GetImportFailedSegmentIDs() ([]int64, error) {
```
comment

EricStarer replied:
what does this mean?

soothing-rain replied:
add comment for this method. But it seems that you commented in the interface so nvm.

EricStarer replied:
ok

soothing-rain commented on internal/rootcoord/import_manager.go:
```diff
@@ -729,3 +729,18 @@ func taskExpired(ti *datapb.ImportTaskInfo) bool {
 }
+
+func (m *importManager) GetImportFailedSegmentIDs() ([]int64, error) {
+	ret := make([]int64, 0, 10)
```
Why is there a cap of 10 and what happens if it is exceeded?

EricStarer replied:
If it's exceeded, the slice will be copy to a new memory space and enlarge its capacity.

soothing-rain replied:
If we not going to limit the size, we can just do: `ret := make([]int64, 0)`.

EricStarer replied:
ok

soothing-rain commented on internal/rootcoord/import_manager.go:
```diff
@@ -729,3 +729,18 @@ func taskExpired(ti *datapb.ImportTaskInfo) bool {
+
+func (m *importManager) GetImportFailedSegmentIDs() ([]int64, error) {
+	ret := make([]int64, 0, 10)
+	for _, importTaskInfo := range m.pendingTasks {
```
locks are required while reading.

soothing-rain commented on internal/datacoord/garbage_collector.go:
```diff
@@ -158,7 +163,33 @@ func (gc *garbageCollector) scan() {
 
 func (gc *garbageCollector) clearEtcd() {
+	failedSegmentIDs := make([]int64, 0)
+	if gc.rcc != nil {
```
Log error and return if rcc == nil

EricStarer replied:
ok

EricStarer replied:
Maybe we can log warn when rcc == nil.

EricStarer commented:
rerun ut

EricStarer commented:
rerun ut
/run-checks

EricStarer commented:
rerun ut

EricStarer commented:
rerun ut

soothing-rain commented on internal/datacoord/garbage_collector.go:
```diff
@@ -158,7 +163,35 @@ func (gc *garbageCollector) scan() {
 	drops := gc.meta.SelectSegments(func(segment *SegmentInfo) bool {
+		_, failedSegmentFlag := failedSegmentIDsSet[segment.ID]
+		if failedSegmentFlag {
+			segment.State = commonpb.SegmentState_Dropped
```
One more thing, do not update segment fields directly. Use copy on write:

https://github.com/milvus-io/milvus/blob/899a9bf86d16bc45422858fa2fff193f043b7c89/internal/datacoord/segment_info.go#L103-L107

EricStarer replied:
done

soothing-rain commented:
/lgtm

EricStarer commented:
/kind improvement

soothing-rain committed:
Refine add segment logic for bulk load

Also restructured `addNewSegment`, `addNormalSegment` and `addFlushedSegment` into a single `addSegment` to avoid duplicate code block and to see more clearly how adding these segments are different.

/kind improvement

Signed-off-by: Yuchen Gao <yuchen.gao@zilliz.com>

soothing-rain commented:
/lgtm
